src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -292,6 +292,31 @@ protected internal void RuntimeLog(byte[] state)
         protected internal void RuntimeNotify(byte[] eventName, Array state)
         {
             if (eventName.Length > MaxEventName) throw new ArgumentException(null, nameof(eventName));
+            HashSet<CompoundType> compounds = new(ReferenceEqualityComparer.Instance);
+            Queue<StackItem> queue = new();
+            queue.Enqueue(state);
+            while (queue.TryDequeue(out var item))
+            {
+                switch (item)
+                {
+                    case Array array:
+                        if (!compounds.Add(array))
+                            throw new InvalidOperationException();
+                        foreach (var subItem in array)
+                            queue.Enqueue(subItem);
+                        break;
+                    case Map map:
+                        if (!compounds.Add(map))
+                            throw new InvalidOperationException();
+                        foreach (var (_, v) in map)
+                            queue.Enqueue(v);
+                        break;
+                    case VM.Types.Buffer:
```

Comment:
I don't think `Buffer` should be forbidden, it's very close to `ByteString` in its essence and depending on the compiler can be used in various circumstances. Notifications are size-limited anyway, so having some 32-byte `Buffer` there doesn't differ much from 32-byte `ByteString`.

Reply:
To execute `Notify` we need to do a `DeepCopy`, to execute `GetNotifications` we need to do a `DeepCopy` again. I think allowing `Buffer` will severely impact performance and lead to attacks. The contract can easily convert `Buffer` to `ByteString` through a `CONVERT` instruction.

Reply:
The copy is only done in `SendNotification()`, so if the buffer is huge we won't get there anyway because of `MaxNotificationSize` limitation. So if any buffer is emitted, it won't be large and I doubt it would make any difference from `ByteString` of the same size. I mostly fear that contracts can emit buffers without knowing it, it all totally depends on the compiler in absence of #1879.

Reply:
You need to `DeepCopy` when executing `GetNotifications`. Otherwise, the contract who received the notifications can modify them. One attack mode is to send a few hundred notifications containing `Buffer`s and then call `GetNotifications` to force automatic `DeepCopy` of all `Buffer`s each time.

Reply:
> You need to DeepCopy when executing GetNotifications

Yep, #1707.

> force automatic DeepCopy of all Buffers

Glancing over VM code I see that `ByteString` is not really copied during `DeepCopy`, so yes, there is a difference. Maybe if we charge per-element in #2748 that'd be irrelevant, though.